internal/worker/sshsession/worker_test.go

```diff
@@ -0,0 +1,315 @@
+	s.facadeClientMock.EXPECT().WatchSSHConnRequest("0").Return(s.watcherMock, nil).AnyTimes()
+
+	// Check the water is Wait()'ed and Kill()'ed exactly once.
+	s.watcherMock.EXPECT().Wait().Times(1)
+	s.watcherMock.EXPECT().Kill().Times(1)
+
+	connGetter, _, _ := newStubConnectionGetter()
+	defer connGetter.close()
+
+	w, err := sshsession.NewWorker(sshsession.WorkerConfig{
+		Logger:               l,
+		MachineId:            "0",
+		FacadeClient:         s.facadeClientMock,
+		ConnectionGetter:     connGetter,
+		EphemeralKeysUpdater: s.ephemeralkeyUpdaterMock,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(workertest.CheckKill(c, w), jc.ErrorIsNil)
+}
+
+// TestSSHSessionWorkerHandlesConnection tests that the worker can at least pipe the
+// connections together using an in-memory net.Pipe. Other than an actual integration
+// test, we cannot test the literal SSH connections to the controller and local SSHD.
+func (s *workerSuite) TestSSHSessionWorkerHandlesConnectionPipesData(c *gc.C) {
+	ctrl := s.setupMocks(c)
+	defer ctrl.Finish()
+
+	l := loggo.GetLogger("test")
+
+	innerChan := make(chan []string)
+	go func() {
+		innerChan <- []string{"machine-0-sshconnectionreq-0"}
+	}()
+	stringChan := watcher.StringsChannel(innerChan)
+
+	s.watcherMock.EXPECT().Wait().AnyTimes()
+	s.watcherMock.EXPECT().Kill().AnyTimes()
+	s.watcherMock.EXPECT().Changes().Return(stringChan).AnyTimes()
+
+	s.facadeClientMock.EXPECT().WatchSSHConnRequest("0").Return(s.watcherMock, nil).Times(1)
+	s.facadeClientMock.EXPECT().GetSSHConnRequest("machine-0-sshconnectionreq-0").Return(
+		params.SSHConnRequest{
+			ControllerAddresses: network.NewSpaceAddresses("127.0.0.1:17022"),
+			EphemeralPublicKey:  []byte{1},
+		},
+		nil,
+	)
+
+	s.ephemeralkeyUpdaterMock.EXPECT().AddEphemeralKey(string([]byte{1})).Times(1)
+	s.ephemeralkeyUpdaterMock.EXPECT().RemoveEphemeralKey(string([]byte{1})).Times(1)
+
+	// Setup an in-memory conn getter to stub the controller and SSHD side.
+	connGetter, controllerConn, sshdConn := newStubConnectionGetter()
+	defer connGetter.close()
+
+	w, err := sshsession.NewWorker(sshsession.WorkerConfig{
+		Logger:               l,
+		MachineId:            "0",
+		FacadeClient:         s.facadeClientMock,
+		ConnectionGetter:     connGetter,
+		EphemeralKeysUpdater: s.ephemeralkeyUpdaterMock,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	defer workertest.CleanKill(c, w)
+
+	go func() {
+		controllerConn.Write([]byte{254})
+		controllerConn.Close()
+	}()
+
+	buf := make([]byte, 1)
+	read, err := sshdConn.Read(buf)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(read, gc.Equals, 1)
+	c.Assert(buf[0], gc.Equals, uint8(254))
+}
+
+// stubConnectionGetter is a stubbed connection getter that returns two in-memory
+// buffered connections. One of a net.Conn representing the sshd connection
+// and the other an ssh.Channel (wrapping a net.Conn) representing the controller
+// connection.
+type stubConnectionGetter struct {
+	sshConn  ssh.Channel
+	sshdConn net.Conn
+}
```

Comment:
There's a lot of boilerplate from here down that I'm pretty sure can be simplified with std-library constructs.